File: apple/InfiniteStreamPlayer/InfiniteStreamPlayer/DeviceInfo.swift
// #550 Phase 4 — device / platform / version taxonomy.
//
// One canonical source for the device fields we stamp on every
// session_events row. Computed lazily once per process; values don't
// change across the app's lifetime so memoising is safe and cheap.
//
// Mirrors the Conviva / Bitmovin breakout: collector identity (the
// "platform" string already in use, e.g. "ios"/"tvos") stays separate
// from hardware class (`device_class`) and physical model
// (`device_model`). `player_tech` names the playback engine so a
// future hls.js / shaka / Roku build can A/B against AVPlayer on the
// same device by varying just that field.

import Foundation
#if canImport(UIKit)
import UIKit
#endif

enum DeviceInfo {
    /// iOS / tvOS / iPadOS major version, e.g. 26 for "26.0.1".
    static let osVersionMajor: UInt16 = parsedOSVersion.major

    /// Minor version, e.g. 0 for "26.0.1". Patch version is dropped
    /// at the schema level (LowCardinality-friendly cardinality).
    static let osVersionMinor: UInt16 = parsedOSVersion.minor

    /// App marketing version from Bundle (CFBundleShortVersionString).
    /// Synced to the repo-root VERSION file by the "Sync version from
    /// repo VERSION" Xcode build phase before every compile — single
    /// source of truth shared with the Android build.gradle, the
    /// server image tag, and the dashboard "What's New" banner. To
    /// bump, edit /VERSION at the repo root; the next build picks
    /// it up automatically and writes both Info.plist files.
    static let appVersion: String = {
        let dict = Bundle.main.infoDictionary
        return (dict?["CFBundleShortVersionString"] as? String) ?? ""
    }()

    /// Form-factor enum: "phone" / "tablet" / "tv" / "desktop" /
    /// "unknown". Derived from UIUserInterfaceIdiom on Apple
    /// platforms; "desktop" reserved for future macOS builds; empty
    /// when UIKit is unavailable.
    static let deviceClass: String = {
        #if canImport(UIKit)
        switch UIDevice.current.userInterfaceIdiom {
        case .phone:        return "phone"
        case .pad:          return "tablet"
        case .tv:           return "tv"
        case .mac:          return "desktop"
        case .carPlay:      return "carplay"
        case .vision:       return "vision"
        case .unspecified:  return "unknown"
        @unknown default:   return "unknown"
        }
        #else
        return "unknown"
        #endif
    }()

    /// Hardware model identifier — e.g. "iPhone15,3" or
    /// "AppleTV14,1". Read via sysctl("hw.machine") because UIDevice
    /// only exposes a generic "iPhone" / "iPad" label.
    static let deviceModel: String = {
        var size = 0
        sysctlbyname("hw.machine", nil, &size, nil, 0)
        guard size > 0 else { return "" }
        var buf = [CChar](repeating: 0, count: size)
        sysctlbyname("hw.machine", &buf, &size, nil, 0)
        return String(cString: buf)
    }()

    /// Playback engine identifier. Hard-coded "AVPlayer" for this
    /// Apple-platform build; a future cross-platform build would set
    /// it to "hls.js" / "shaka" / "native-roku" etc.
    static let playerTech: String = "AVPlayer"

    /// Physical-pixel resolution of the device's current orientation,
    /// formatted as `"WxH"` to match `video_resolution` /
    /// `display_resolution` for side-by-side comparison. On iPad this
    /// swaps when the device rotates; on Apple TV / iPhone-locked
    /// orientations it stays constant. Computed each call so the
    /// caller gets the current orientation rather than the value at
    /// app launch.
    ///
    /// Replaces the prior `screen_width_px` / `screen_height_px` /
    /// `screen_density` taxonomy fields, which were static
    /// portrait-only readings and never reflected rotation.
    static func deviceResolution() -> String {
        #if canImport(UIKit)
        let b = UIScreen.main.bounds
        let s = UIScreen.main.nativeScale
        let w = Int((b.width  * s).rounded())
        let h = Int((b.height * s).rounded())
        guard w > 0 && h > 0 else { return "" }
        return "\(w)x\(h)"
        #else
        return ""
        #endif
    }

    // ── Helpers ────────────────────────────────────────────────────

    private static let parsedOSVersion: (major: UInt16, minor: UInt16) = {
        let v = ProcessInfo.processInfo.operatingSystemVersion
        let major = (v.majorVersion >= 0 && v.majorVersion <= Int(UInt16.max))
            ? UInt16(v.majorVersion) : 0
        let minor = (v.minorVersion >= 0 && v.minorVersion <= Int(UInt16.max))
            ? UInt16(v.minorVersion) : 0
        return (major, minor)
    }()
}

File: android/InfiniteStreamPlayer/app/src/main/java/com/infinitestream/player/DeviceInfo.java
package com.infinitestream.player;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Build;
import android.util.DisplayMetrics;

import androidx.media3.common.MediaLibraryInfo;

// #550 Phase 4 — device / platform / version taxonomy.
//
// Android counterpart to the Swift DeviceInfo in the Apple build.
// One canonical source for the device fields we stamp on every
// session_events row. Values don't change across the process's
// lifetime so memoising on first read is safe and cheap.
//
// Mirrors the Conviva / Bitmovin breakout: collector identity
// (player_metrics_source already in use, e.g. "android") stays
// separate from hardware class (`device_class`) and physical model
// (`device_model`). `player_tech` names the playback engine so a
// future Cast / hls.js Android build can A/B against ExoPlayer on
// the same device by varying just that field.
public final class DeviceInfo {

    private DeviceInfo() { }

    /** OS major version, e.g. 14 for Android 14 (API 34). */
    public static int osVersionMajor() {
        return parseFirstInt(Build.VERSION.RELEASE, 0);
    }

    /** OS minor version, e.g. 0 for "14.0". Patch dropped at the
     *  schema level (LowCardinality-friendly cardinality). */
    public static int osVersionMinor() {
        String r = Build.VERSION.RELEASE;
        if (r == null) return 0;
        int dot = r.indexOf('.');
        if (dot < 0 || dot + 1 >= r.length()) return 0;
        return parseFirstInt(r.substring(dot + 1), 0);
    }

    /** App marketing version from the gradle build script (synced to
     *  the repo-root VERSION file). Single source of truth shared with
     *  the iOS DeviceInfo, the server image tag, and the dashboard
     *  "What's New" banner. */
    public static String appVersion() {
        return BuildConfig.VERSION_NAME == null ? "" : BuildConfig.VERSION_NAME;
    }

    /** Form-factor enum: "phone" / "tablet" / "tv". Derived from
     *  Configuration.uiMode + screen size. */
    public static String deviceClass(Context ctx) {
        if (ctx == null) return "unknown";
        int mode = ctx.getResources().getConfiguration().uiMode
            & Configuration.UI_MODE_TYPE_MASK;
        if (mode == Configuration.UI_MODE_TYPE_TELEVISION) return "tv";
        int sizeBits = ctx.getResources().getConfiguration().screenLayout
            & Configuration.SCREENLAYOUT_SIZE_MASK;
        if (sizeBits >= Configuration.SCREENLAYOUT_SIZE_LARGE) return "tablet";
        return "phone";
    }

    /** Hardware model identifier, e.g. "Pixel 7" or "AFTKA". Read
     *  from android.os.Build.MODEL. */
    public static String deviceModel() {
        return Build.MODEL == null ? "" : Build.MODEL;
    }

    /** Playback engine identifier. Hard-coded "ExoPlayer" for this
     *  Media3-based build. */
    public static String playerTech() {
        return "ExoPlayer";
    }

    /** Playback engine VERSION — the Media3/ExoPlayer library version,
     *  e.g. "1.2.1". Read at runtime from MediaLibraryInfo so it tracks
     *  the bundled dependency automatically (no hardcoding, survives a
     *  gradle bump). Crucially this is INDEPENDENT of the OS version:
     *  ExoPlayer is an app-bundled library, so two devices on the same
     *  Android release can run different player versions. (Contrast iOS,
     *  where AVPlayer IS the OS and the OS version is its version.)
     *  Pairs with player_tech for per-engine-version A/B + regression
     *  attribution. */
    public static String playerTechVersion() {
        return MediaLibraryInfo.VERSION == null ? "" : MediaLibraryInfo.VERSION;
    }

    /** Orientation-aware physical pixels of the device screen, "WxH".
     *  Mirrors iOS's `device_resolution`: changes on rotation (W and H
     *  swap) but reflects hardware capability, not the playback view's
     *  render surface. The render-surface counterpart is
     *  player_metrics_display_resolution stamped from PlayerView dims.
     *  Replaces the prior screen_width_px / screen_height_px / screen_density
     *  trio so the schema stays a single tile per concept. */
    public static String deviceResolution(Context ctx) {
        if (ctx == null) return "";
        DisplayMetrics dm = ctx.getResources().getDisplayMetrics();
        if (dm.widthPixels <= 0 || dm.heightPixels <= 0) return "";
        return dm.widthPixels + "x" + dm.heightPixels;
    }

    // Read leading run of digits as an int. Tolerates trailing
    // strings like "14.0" or "13-rc1". Falls back to defaultValue
    // if no leading digit is present.
    private static int parseFirstInt(String s, int defaultValue) {
        if (s == null) return defaultValue;
        int n = 0;
        boolean any = false;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c < '0' || c > '9') break;
            n = n * 10 + (c - '0');
            any = true;
        }
        return any ? n : defaultValue;
    }
}
